### PR TITLE
fix: dir package isolates system/user directory

### LINKED ISSUE #144: directory structure needs to isolate `read` and `write` path
As current design, when system level directory has the config file while user level directory doesn't have the config file, it uses the system level config file for read and write. 

It is necessary to allow the user to know that whether the write path is user level or system level, so we need to:
1. isolate `read` /`write` path
2. provide the ability to select `user`/`system` level directory to write.
Linked to [System vs User file/directory precedence#203](https://github.com/notaryproject/notation/issues/203)

### PR DESCRIPTION
**Pull request by JeyJeyGao**
1. config add EnvelopeType argument.
2. dir package optimize: allow user define DirLevel = UserLevel | SystemLevel | UnionLevel.
4. refector config pacakge: receive a path argument when load and save file to allow user to overwrite the default directory structure.

Test: added unit test.

Resolves #144 
Signed-off-by: Junjie Gao <junjiegao@microsoft.com>

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -37,15 +41,30 @@ func NewConfig() *Config {
 	}
 }
 
-// Save stores the config to file
-func (c *Config) Save() error {
-	return save(ConfigPath, c)
+// Save stores the config to file.
+//
+// if the `path` is an empty string, it uses build-in user level config directory.
```
[the remark this answers is not in the document]
Comment (JeyJeyGao) [reply config/config.go]: Thanks!

File: config/config.go [hunk 2]
```diff
@@ -37,15 +41,30 @@ func NewConfig() *Config {
...
 }
 
 // LoadConfig reads the config from file or return a default config if not found.
-func LoadConfig() (*Config, error) {
+//
+// if `path` is an empty string, it uses build-in config.json directory, including
+// user level and system level.
+func LoadConfig(path string) (*Config, error) {
+	// set default path
+	if path == "" {
```
[the remark this answers is not in the document]
Comment (JeyJeyGao) [reply config/config.go:61]: If `path` is empty, it uses the default value. It is not elegant to provide a default value by using an empty `path` string, but Golang doesn't support the default function parameter. If I support more invalid `path` string, the code will be much not elegant. In order to use the default value, different user will call it by different invalid `path` sring.

**Comment by patrickzheng200** [comment]
Looks like we can move the target of this PR to the dev-rc.1 branch for now?

**Comment by JeyJeyGao** [comment]
>

**Comment by JeyJeyGao** [comment]
> Looks like we can move the target of this PR to the dev-rc.1 branch for now?

This PR has no dependent on COSE refactor, so it can be merged to main branch.

File: dir/path.go [hunk 3]
```diff
@@ -36,13 +36,28 @@ const (
...
+// WriteLevel defines the write directory level supporting UserLevel or SystemLevel.
+type WriteLevel int
+
+const (
+	// SystemLevel is the label to specify write directory to system level
+	SystemLevel WriteLevel = 0
+	// UserLevel is the label to specify write directory to user level
+	UserLevel WriteLevel = 1
+)
+
```
Comment (priteshbandi) [inline dir/path.go:48]: Why levels are specific to write? can we make them generic so they be used for read too?

File: config/config.go [hunk 4]
```diff
@@ -23,6 +25,8 @@ type Config struct {
 	InsecureRegistries       []string                 `json:"insecureRegistries"`
 	CredentialsStore         string                   `json:"credsStore,omitempty"`
 	CredentialHelpers        map[string]string        `json:"credHelpers,omitempty"`
+	// EnvelopeType defines the envelope type for signing
+	EnvelopeType string `json:"envelopeType,omitempty"`
```
Comment (priteshbandi) [inline config/config.go:29]: This change is out of scope for https://github.com/notaryproject/notation/issues/203, please remove it from this PR.

File: dir/base.go [hunk 5]
```diff
@@ -93,12 +93,21 @@ func loadPath() {
 		UserConfigFS: NewUnionDirFS(
 			NewRootedFS(userConfig, nil),
 		),
+		SystemConfigFS: NewUnionDirFS(
+			NewRootedFS(userConfig, nil),
+		),
```
Comment (priteshbandi) [inline dir/base.go:98]: Why do we need `SystemConfigFS` and `UserConfigFS` when we already have `ConfigFS`? I am assuming because `ConfigFS` tries to tries to read the user dir first then the system dir? If so, it raises the question why do we need `fs.go`? Can we get rid of it and the complications that come with it and rewrite it as simple utility methods which return paths based on the config we define(in spec)

File: dir/base.go [hunk 6]
```diff
@@ -93,12 +93,21 @@ func loadPath() {
...
 		LibexecFS: NewUnionDirFS(
 			NewRootedFS(userLibexec, nil),
 			NewRootedFS(systemLibexec, nil),
 		),
+		UserLibexecFS: NewUnionDirFS(
+			NewRootedFS(userLibexec, nil),
+		),
+		SystemLibexecFS: NewUnionDirFS(
+			NewRootedFS(systemLibexec, nil),
+		),
```
Comment (priteshbandi) [inline dir/base.go:111]: ditto

File: dir/path.go [hunk 7]
```diff
@@ -69,27 +89,38 @@ func (p *PathManager) Localkey(name string) (keyPath, certPath string) {
...
 	return path
 }
 
-// TrustPolicy returns the path of trustpolicy.json file
+// TrustPolicy returns the ready-only path of trustpolicy.json file
 func (p *PathManager) TrustPolicy() string {
 	path, err := p.ConfigFS.GetPath(TrustPolicyFile)
 	checkError(err)
 	return path
 }
```
Comment (priteshbandi) [inline dir/path.go:104]: Instead of creating a separate function for reading and writing, can we write a function that takes UserLevel/SystemLevel as a param and determines the path internally?

File: dir/path.go [hunk 8]
```diff
@@ -69,27 +89,38 @@ func (p *PathManager) Localkey(name string) (keyPath, certPath string) {
...
 func (p *PathManager) SigningKeyConfig() string {
 	path, err := p.UserConfigFS.GetPath(SigningKeysFile)
 	checkError(err)
 	return path
 }
 
-// TrustPolicy returns the path of trustpolicy.json file
+// TrustPolicy returns the ready-only path of trustpolicy.json file
 func (p *PathManager) TrustPolicy() string {
 	path, err := p.ConfigFS.GetPath(TrustPolicyFile)
```
Comment (priteshbandi) [inline dir/path.go:101]: The path of trystpolicy file is know(based on system or user level) then why are we searching that in directory using `GetPath`?

File: config/config.go [hunk 9]
```diff
@@ -37,15 +41,30 @@ func NewConfig() *Config {
...
-func (c *Config) Save() error {
-	return save(ConfigPath, c)
+// Save stores the config to file.
+//
+// if the `path` is an empty string, it uses built-in user level config directory.
+func (c *Config) Save(path string) error {
+	// set default path
+	if path == "" {
+		path = dir.Path.ConfigForWrite(dir.UserLevel)
+	}
```
Comment (priteshbandi) [inline config/config.go:51]: Why do we need this if can we take this path from caller? 
Applicable for all load and save functions in config.go and keys.go.

File: dir/path.go [hunk 10]
```diff
@@ -52,14 +67,19 @@ func checkError(err error) {
...
+// Config returns the ready-only path of config.json
 func (p *PathManager) Config() string {
 	path, err := p.ConfigFS.GetPath(ConfigFile)
 	checkError(err)
 	return path
 }
 
-// LocalKey returns path of the local private key and it's certificate
+// ConfigForWrite returns the writable path of config.json
+func (p *PathManager) ConfigForWrite(writeLevel WriteLevel) string {
```
Comment (priteshbandi) [inline dir/path.go:78]: Instead of creating a separate function for reading and writing, can we write a function that takes UserLevel/SystemLevel as a param and determines the path internally?

Also, applicable to other similar functions.

**Review by rgnote — COMMENTED** [review]
Is there a spec update for adding `EnvelopeType`? If yes, can you provide that link?

**Review by priteshbandi — COMMENTED** [review]
Left some feedback on overall structure of config module. It looks like the `UnionDirFS` abstraction is making it complex to add system level and user level read/write permutations .

-------

Also, depending on decision we make in https://github.com/notaryproject/notation/issues/203, we might need to make significant changes in this PR.

File: config/config.go (hunk 4 above)
Comment (JeyJeyGao) [reply config/config.go]: removed.

File: dir/base.go (hunk 5 above)
Comment (JeyJeyGao) [reply dir/base.go:98]: The reason that I uses io/fs package is `./plugin/manager/manager.go` uses io/fs to WalkDir the directory to list all available plugins. According to https://github.com/notaryproject/notation/blob/main/specs/directory.md, if both system and user level have plugins, it should walk through both of them to find plugins. The UnionDirFS in fs.go provides a fs.FS to read multiple directories, so the [manager.go](https://github.com/notaryproject/notation-go/blob/main/plugin/manager/manager.go#L91) can use it. 

The UnionDirFS can read multiple directories with priority, however, it can not distinguish user or system directory, so I add 3 FS to cover different use cases.
```
                // need to walk through both user and system level
		ConfigFS: NewUnionDirFS(
			NewRootedFS(userConfig, nil),
			NewRootedFS(systemConfig, nil),
		),
                // write to user level
		UserConfigFS: NewUnionDirFS(
			NewRootedFS(userConfig, nil),
		),
                // write to system level
		SystemConfigFS: NewUnionDirFS(
			NewRootedFS(userConfig, nil),
		),
```

File: dir/path.go (hunk 8 above)
Comment (JeyJeyGao) [reply dir/path.go:101]: GetPath will search both `user` and `system` level, and user level has higher priority, so the GetPath helps search and decide the priority.

File: config/config.go (hunk 9 above)
Comment (JeyJeyGao) [reply config/config.go:49]: Because we have directory spec and it should be the default case (an empty path), and it is convenient for user to not specify the directory.

File: dir/path.go (hunk 10 above)
Comment (JeyJeyGao) [reply dir/path.go]: Sure, updated.
Also, renamed `WriteLevel` to `DirLevel`, and added `UnionLevel`

File: dir/path.go (hunk 7 above)
Comment (JeyJeyGao) [reply dir/path.go]: Updated.

File: dir/base.go (hunk 5 above)
Comment (JeyJeyGao) [reply dir/base.go]: Removed `UserConfigFS` and `SystemConfigFS`, please review it again.

File: dir/base.go (hunk 6 above)
Comment (JeyJeyGao) [reply dir/base.go]: Removed. Please review it again.

File: dir/path.go (hunk 3 above)
Comment (JeyJeyGao) [reply dir/path.go]: Renamed `WriteLevel` to `DirLevel`

**Comment by JeyJeyGao** [comment]
> Added some feedback on overall structure of config module. It looks like the `UnionDirFS` abstraction is making it complex to add system level and user level read/write permutations, we should rethink about it .
> 
> Also, depending on decision we make in [notaryproject/notation#203](https://github.com/notaryproject/notation/issues/203), we might need to make significant changes in this PR.

Refactor it again. Now, `UnionDirFS` is only used to GetPath() in both system and user directory. If user specify the System/User level directory, return the path directly.

File: config/config.go (hunk 9 above)
Comment (JeyJeyGao) [reply config/config.go]: Sure. Deleted.

**Comment by priteshbandi** [comment]
We can close/update this PR and update/open new one for only supporting user level config

https://github.com/notaryproject/notation/issues/203#issuecomment-1281663886
